<a href="https://colab.research.google.com/github/jayounghoyos/An-lisis-de-Datos-con-Pandas-y-Python/blob/main/Data_acquisition_%26_data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data acquisition & data cleaning

Integrantes:
- Juan Andrés Young Hoyos
- Martín Valencia Vallejo
- María Acevedo Suárez

In [81]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [82]:
!ls '/content/drive/MyDrive/FundamentosDeAprendizajeEAFIT'

Titanic-Dataset.csv


# Nuestra exploración de los datos

In [83]:
dataset = pd.read_csv('/content/drive/MyDrive/FundamentosDeAprendizajeEAFIT/Titanic-Dataset.csv')
dataset

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [84]:
dataset.shape

(891, 12)

In [85]:
dataset.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [86]:
dataset.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [87]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [88]:
dataset.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [89]:
dataset.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [90]:
dataset.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [91]:
dataset.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [92]:
dataset.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [93]:
dataset.duplicated().sum()

np.int64(0)

In [94]:
dataset['Fare'].value_counts()

Fare
8.0500     43
13.0000    42
7.8958     38
7.7500     34
26.0000    31
           ..
13.8583     1
50.4958     1
5.0000      1
9.8458      1
10.5167     1
Name: count, Length: 248, dtype: int64

### Limpieza de los ***datos***

In [95]:
dataset.drop("PassengerId",axis=1)
dataset.drop("Name", axis=1)
dataset.drop("Ticket", axis=1)

dataset.head()
dataset.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Este código realiza pasos esenciales de preprocesamiento de datos en el conjunto de datos de Titanic para prepararlo para construir un modelo de IA que prediga la supervivencia de los pasajeros.

1. Carga del conjunto de datos:
  - El conjunto de datos se carga desde un archivo CSV en un DataFrame de pandas, lo que permite una fácil manipulación y análisis de datos.

2. Eliminación de columnas irrelevantes o redundantes:
  - 'PassengerId': Un identificador único para cada pasajero, que no proporciona información predictiva.
  - 'Ticket': Los números de ticket son en su mayoría arbitrarios y es poco probable que ayuden al modelo a aprender patrones de supervivencia.
  - 'Name': Los nombres son únicos y textuales; sin un procesamiento adicional (como la extracción de títulos), no contribuyen directamente a la predicción.
  - 'Cabin': Esta columna tiene muchos valores faltantes y datos inconsistentes, lo que la hace poco confiable sin una imputación compleja.
  - 'Embarked': Aunque podría tener algún valor predictivo, se elimina aquí, posiblemente para simplificar el modelo o debido a valores faltantes.
  - 'Fare': La tarifa podría correlacionarse con la supervivencia, pero se excluye aquí, probablemente para centrarse en otras características o debido a consideraciones de calidad de los datos. La eliminación de estas columnas reduce el ruido y la dimensionalidad, ayudando al modelo a centrarse en características más relevantes.


# Mínima limpieza con base en el análisis

In [96]:
#Authors: Martín Valencia, María Acevedo, Juan Andrés Young

import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/FundamentosDeAprendizajeEAFIT/Titanic-Dataset.csv')


'''
Impresión del DataFrame limpio:
  - Este paso permite verificar que los datos se hayan limpiado correctamente, con las columnas previstas eliminadas y los valores faltantes manejados.
'''

df = df.drop("Ticket", axis = 1)
df = df.drop("PassengerId", axis = 1)
df = df.drop("Embarked", axis = 1)
df = df.drop("Fare", axis = 1)
df = df.drop("Name", axis = 1)
df = df.drop("Cabin", axis = 1)

print(df)

     Survived  Pclass     Sex   Age  SibSp  Parch
0           0       3    male  22.0      1      0
1           1       1  female  38.0      1      0
2           1       3  female  26.0      0      0
3           1       1  female  35.0      1      0
4           0       3    male  35.0      0      0
..        ...     ...     ...   ...    ...    ...
886         0       2    male  27.0      0      0
887         1       1  female  19.0      0      0
888         0       3  female   NaN      1      2
889         1       1    male  26.0      0      0
890         0       3    male  32.0      0      0

[891 rows x 6 columns]


# Máxima limpieza ahora con un análisis profundo de las edades y como se trata con esos datos

In [97]:
df = df.fillna({"Age" : df["Age"].mean()})

'''
Manejo de valores faltantes en la columna 'Age':
  - La característica 'Age' es importante porque la edad puede influir en las posibilidades de supervivencia (por ejemplo, los niños podrían haber sido priorizados).
  - Los valores de edad faltantes se rellenan con la edad media de todos los pasajeros, una técnica de imputación común y simple.
  - Esto asegura que no queden valores faltantes, lo cual es necesario porque muchos algoritmos de aprendizaje automático no pueden manejar datos faltantes. '''

print(df)

     Survived  Pclass     Sex        Age  SibSp  Parch
0           0       3    male  22.000000      1      0
1           1       1  female  38.000000      1      0
2           1       3  female  26.000000      0      0
3           1       1  female  35.000000      1      0
4           0       3    male  35.000000      0      0
..        ...     ...     ...        ...    ...    ...
886         0       2    male  27.000000      0      0
887         1       1  female  19.000000      0      0
888         0       3  female  29.699118      1      2
889         1       1    male  26.000000      0      0
890         0       3    male  32.000000      0      0

[891 rows x 6 columns]


# Exportar datasets para aplicarModeloML.ipynb
## 1.3 Comparacion entre limpieza minima y maxima

### 1.3.1 Limpieza minima
Solo lo necesario para que el algoritmo de ML corra sin errores:
- Eliminar columnas no numericas (Name, Sex, Ticket, Cabin, Embarked) para evitar errores de tipo
- Eliminar filas con NaN (Age) para evitar errores de valores faltantes
- Se pierde informacion valiosa (Sex es el predictor #1 del Titanic) y se pierden 177 filas

In [98]:
import pandas as pd

df_min = pd.read_csv('/content/drive/MyDrive/FundamentosDeAprendizajeEAFIT/Titanic-Dataset.csv')

# Solo eliminar columnas no numericas y PassengerId
df_min = df_min.drop(columns=['PassengerId', 'Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'])

# Eliminar filas con NaN para que no haya errores
df_min = df_min.dropna()

df_min.to_csv('titanic_min.csv', index=False)
print(f'Shape limpieza minima: {df_min.shape}')
print(f'Columnas: {list(df_min.columns)}')
df_min.head()

Shape limpieza minima: (714, 6)
Columnas: ['Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']


Survived  Pclass   Age  SibSp  Parch     Fare
0         0       3  22.0      1      0   7.2500
1         1       1  38.0      1      0  71.2833
2         1       3  26.0      0      0   7.9250
3         1       1  35.0      1      0  53.1000
4         0       3  35.0      0      0   8.0500

### 1.3.2 Limpieza maxima
Limpieza exhaustiva para maximizar el desempenio del modelo:
- Imputar Age con la mediana por Pclass (mas preciso que media global)
- Imputar Embarked con la moda
- Codificar Sex con LabelEncoder (female=0, male=1)
- One-Hot Encoding para Embarked
- Feature engineering: FamilySize y IsAlone
- Se conservan todas las 891 filas y se agregan features informativos

In [99]:
from sklearn.preprocessing import LabelEncoder

df_max = pd.read_csv('/content/drive/MyDrive/FundamentosDeAprendizajeEAFIT/Titanic-Dataset.csv')

# Eliminar columnas irrelevantes
df_max = df_max.drop(columns=['PassengerId', 'Name', 'Ticket', 'Cabin'])

# Imputar Age con mediana por Pclass
df_max['Age'] = df_max.groupby('Pclass')['Age'].transform(lambda x: x.fillna(x.median()))

# Imputar Embarked con la moda
df_max['Embarked'] = df_max['Embarked'].fillna(df_max['Embarked'].mode()[0])

# Codificar Sex
le = LabelEncoder()
df_max['Sex'] = le.fit_transform(df_max['Sex'])

# One-Hot Encoding para Embarked
df_max = pd.get_dummies(df_max, columns=['Embarked'], drop_first=True)

# Feature engineering
df_max['FamilySize'] = df_max['SibSp'] + df_max['Parch'] + 1
df_max['IsAlone'] = (df_max['FamilySize'] == 1).astype(int)

# Convertir columnas bool a int (por compatibilidad)
for col in df_max.select_dtypes(include='bool').columns:
    df_max[col] = df_max[col].astype(int)

df_max.to_csv('titanic_max.csv', index=False)
print(f'Shape limpieza maxima: {df_max.shape}')
print(f'Columnas: {list(df_max.columns)}')
print(f'Valores faltantes: {df_max.isnull().sum().sum()}')
df_max.head()

Shape limpieza maxima: (891, 11)
Columnas: ['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked_Q', 'Embarked_S', 'FamilySize', 'IsAlone']
Valores faltantes: 0


Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_Q  Embarked_S  \
0         0       3    1  22.0      1      0   7.2500           0           1   
1         1       1    0  38.0      1      0  71.2833           0           0   
2         1       3    0  26.0      0      0   7.9250           0           1   
3         1       1    0  35.0      1      0  53.1000           0           1   
4         0       3    1  35.0      0      0   8.0500           0           1   

   FamilySize  IsAlone  
0           2        0  
1           2        0  
2           1        1  
3           2        0  
4           1        1

---
## Aplicar Modelo de ML (aplicarModeloML.ipynb)
Algoritmo: **RandomForestClassifier** (clasificacion binaria: Survived 0/1)

Se usa el mismo algoritmo con los mismos parametros en ambas limpiezas para comparar el efecto de la limpieza en el desempenio.

In [100]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

### Modelo con limpieza minima

In [101]:
# PREPARACION
X_min = df_min.drop(columns=['Survived'])
y_min = df_min['Survived']

X_train_min, X_test_min, y_train_min, y_test_min = train_test_split(
    X_min, y_min, test_size=0.2, random_state=42
)

# ESCALAMIENTO
scaler_min = StandardScaler()
X_train_min = scaler_min.fit_transform(X_train_min)
X_test_min = scaler_min.transform(X_test_min)

# ENTRENAMIENTO CON RANDOM FOREST
algorithm_min = RandomForestClassifier(n_estimators=100, random_state=42)
algorithm_min.fit(X_train_min, y_train_min)

# PREDICCION Y ACCURACY
y_pred_min = algorithm_min.predict(X_test_min)
acc_min = accuracy_score(y_test_min, y_pred_min)

print(f"--- RESULTADOS LIMPIEZA MINIMA ---")
print(f"Accuracy Global: {acc_min * 100:.2f}%")
print("-" * 33)
print(f"\nClassification Report:")
print(classification_report(y_test_min, y_pred_min))
print(f"Confusion Matrix:")
print(confusion_matrix(y_test_min, y_pred_min))

--- RESULTADOS LIMPIEZA MINIMA ---
Accuracy Global: 68.53%
---------------------------------

Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.82      0.76        87
           1       0.63      0.48      0.55        56

    accuracy                           0.69       143
   macro avg       0.67      0.65      0.65       143
weighted avg       0.68      0.69      0.68       143

Confusion Matrix:
[[71 16]
 [29 27]]


### Modelo con limpieza maxima

In [102]:
# PREPARACION
X_max = df_max.drop(columns=['Survived'])
y_max = df_max['Survived']

X_train_max, X_test_max, y_train_max, y_test_max = train_test_split(
    X_max, y_max, test_size=0.2, random_state=42
)

# ESCALAMIENTO
scaler_max = StandardScaler()
X_train_max = scaler_max.fit_transform(X_train_max)
X_test_max = scaler_max.transform(X_test_max)

# ENTRENAMIENTO CON RANDOM FOREST (mismos parametros)
algorithm_max = RandomForestClassifier(n_estimators=100, random_state=42)
algorithm_max.fit(X_train_max, y_train_max)

# PREDICCION Y ACCURACY
y_pred_max = algorithm_max.predict(X_test_max)
acc_max = accuracy_score(y_test_max, y_pred_max)

print(f"--- RESULTADOS LIMPIEZA MAXIMA ---")
print(f"Accuracy Global: {acc_max * 100:.2f}%")
print("-" * 33)
print(f"\nClassification Report:")
print(classification_report(y_test_max, y_pred_max))
print(f"Confusion Matrix:")
print(confusion_matrix(y_test_max, y_pred_max))

--- RESULTADOS LIMPIEZA MAXIMA ---
Accuracy Global: 82.12%
---------------------------------

Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.86      0.85       105
           1       0.79      0.77      0.78        74

    accuracy                           0.82       179
   macro avg       0.82      0.81      0.81       179
weighted avg       0.82      0.82      0.82       179

Confusion Matrix:
[[90 15]
 [17 57]]


### Comparacion de resultados

In [103]:
print("=" * 40)
print("       COMPARACION DE RESULTADOS")
print("=" * 40)
print(f"Accuracy limpieza minima:  {acc_min * 100:.2f}%")
print(f"Accuracy limpieza maxima:  {acc_max * 100:.2f}%")
print(f"Mejora:                    {(acc_max - acc_min) * 100:.2f}%")
print("=" * 40)

       COMPARACION DE RESULTADOS
Accuracy limpieza minima:  68.53%
Accuracy limpieza maxima:  82.12%
Mejora:                    13.59%


### Conclusiones

- **Limpieza minima**: Se usaron solo columnas numericas originales (Pclass, Age, SibSp, Parch, Fare), se eliminaron 177 filas con NaN en Age, y se descarto informacion clave como Sex (el predictor mas fuerte de supervivencia en el Titanic).

- **Limpieza maxima**: Se conservaron las 891 filas al imputar valores faltantes, se codifico Sex y Embarked, y se crearon features nuevos (FamilySize, IsAlone). El modelo tuvo acceso a mas informacion relevante.

- La limpieza maxima produce un **mejor desempenio** porque el modelo tiene acceso a Sex (women and children first), mas datos de entrenamiento, y features adicionales que capturan patrones de supervivencia.